### PR TITLE
Use `configure-aws-credentials` workflow instead of passing `secret_access_key`

### LINKED ISSUE #1346: Use `configure-aws-credentials` instead of passing `secret_access_key`
**The problem/use-case that the feature addresses**
API calls to AWS need to be signed with credential information, so when you use one of the AWS SDKs or an AWS tool, you must provide it with AWS credentials and and AWS region. One way to do that in GitHub Actions is to use a repository secret with IAM credentials, but this doesn't follow [AWS security guidelines](https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html) on using long term creds.

**Description of the feature**
* First we need to set up OpenID Connect. [OpenID with AWS](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services)
* Then Implement the workflow using [configure-aws-credentials](https://github.com/aws-actions/configure-aws-credentials?tab=readme-ov-file)

```
- name: Configure AWS Credentials for China region audience
      uses: aws-actions/configure-aws-credentials@v4
      with:
        aws-region: us-east-3
        role-to-assume: arn:aws-cn:iam::123456789100:role/my-github-actions-role
```
By using this approach we can get rid off the secret keys needed to pass in and also organize the workflow

### PR DESCRIPTION
**Pull request by vudiep411**
## Summary
This PR fixes #1346 where we can get rid of the long term credentials by using OpenID Connect. OpenID Connect (OIDC) allows your GitHub Actions workflows to access resources in Amazon Web Services (AWS), without needing to store the AWS credentials as long-lived GitHub secrets.

## Changes
We can remove these secrets that were passed in previously:
```
token: ${{ secrets.GITHUB_TOKEN }}
      bucket: ${{ secrets.AWS_S3_BUCKET }}
      access_key_id: ${{ secrets.AWS_S3_ACCESS_KEY_ID }}
      secret_access_key: ${{ secrets.AWS_S3_ACCESS_KEY }}
```

Instead we only need the `role-to-assume` arn. For more information [OIDC](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect).

## Prerequisites
Before merging this PR, we need to make sure to set up the proper Identity providers on the production AWS account. Follow this [guides](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services). 
 **Quick guide**:
1. Create an Identity provider with OpenID Connect.
Provider url: `https://token.actions.githubusercontent.com` <br/>
Audience: `sts.amazonaws.com`
Screenshot 2024-11-26 at 1 02 23 PM

2. Assign the role into this new provider with this trust policy like below
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::<aws_account_id>:oidc-provider/token.actions.githubusercontent.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringLike": {
                    "token.actions.githubusercontent.com:sub": [
                        "repo:valkey-io/valkey:ref:refs/heads/unstable",
                        "repo:valkey-io/valkey:ref:refs/tags/*"
                    ],
                    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```
3. In Github secrets, Add these secrets:
Screenshot 2024-11-26 at 2 11 02 PM

## Results
**Github action run:**
Screenshot 2024-11-26 at 2 06 17 PM

**Files in S3 Dev env:**
Screenshot 2024-11-26 at 2 09 42 PM

> Note: the failed workflow can be found in this issue #1343 which will be in a separate PR.

### CODE REVIEW

**Review by enjoy-binbin — APPROVED** [review]
seems ok, @roshkhatri please take a look

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml:19]: Why do we need the write permissions here?

File: .github/workflows/call-build-linux-arm-packages.yml
Comment (roshkhatri) [inline .github/workflows/call-build-linux-arm-packages.yml:31]: why do we need write permissions here?

File: .github/workflows/call-build-linux-x86-packages.yml
Comment (roshkhatri) [inline .github/workflows/call-build-linux-x86-packages.yml:69]: Is the `aws cli` installed by `aws-actions/configure-aws-credentials@v4` ??

**Review by roshkhatri — COMMENTED** [review]
LGTM, just had a few questions.

**Comment by roshkhatri** [comment]
We will also need to do the prerequisites steps for the main repo before we merge this.

File: .github/workflows/build-release-packages.yml
Comment (vudiep411) [reply .github/workflows/build-release-packages.yml:19]: This is to ensure that your workflow has the necessary permission to perform OIDC authentication with AWS.

File: .github/workflows/call-build-linux-arm-packages.yml
Comment (vudiep411) [reply .github/workflows/call-build-linux-arm-packages.yml:31]: Same as the comment above. Each job spawned in the matrix runs in a separate environment and each of this env needs permission to perform OIDC authentication with AWS.

File: .github/workflows/call-build-linux-x86-packages.yml
Comment (vudiep411) [reply .github/workflows/call-build-linux-x86-packages.yml:69]: Yes, `aws-actions/configure-aws-credentials@v4` will configured the runner environment with the proper AWS credentials and CLI

**Comment by roshkhatri** [comment]
@vudiep411, will it be possible to add a test, where we uploads a test build binary to a test s3 bucket when ever changes are made to these workflows.

With this we can be sure that when we release stuff, it doesn't break on the main valkey repository

**Comment by vudiep411** [comment]
Yes it is possible, we can use github environment for that. So we can write something like this:
```
- name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v2
        with:
          role-to-assume: ${{ env.ROLE_TO_ASSUME }}
          role-session-name: ${{ env.ROLE_SESSION_NAME }}
          aws-region: ${{ env.AWS_REGION }}
```
We use env to dynamically assigned the role to assume and env. But this would required us to set up multiple OCID on different accounts not just in the workflow itself so it should be in a separate issue. For now we can just set this one up and after this looks good, I will open another issue for it if that's ok.
@roshkhatri

**Comment by roshkhatri** [comment]
I think we can also do it based on the github event trigger maybe? 
1. We set two OCID for two buckets, `releases` and `test`.
2. When the changes are made to release workflow files, for a `pull_request` event we run the same workflow with `secrets.AWS_TEST_ROLE_TO_ASSUME` and secrets.AWS_TEST_S3_BUCKET and for a `workflow_dispatch` and `release`
 event trigger we run with the current setup.
3. We would do it at the head of unstable.
4. Once we upload the test binaries, we can confirm the upload, delete the delete the file and pass the test.

Also, I think we should add the test on the same PR so we can run the test on this PR and know that the tests also work. We would also have to do the pre-requisites only once for both the scenarios.

**Comment by vudiep411** [comment]
That's a really good suggestion. We can look into that. That way it would automated the testing of a PR with a test bucket. Definitely doable

File: .github/workflows/build-release-packages.yml
Comment (vudiep411) [inline .github/workflows/build-release-packages.yml:80]: We can use version output from `release-build-get-meta`

**Comment by vudiep411** [comment]
## New Changes
1. When a PR is raised, the build-release will run if there are changes to the release workflow or `utils/releasetools`
2. `Get the version` step will output the version that is getting built. eg. 8.0.0, unstable,...
3. `Set bucket name` test will output **true** if the event is a PR else it will output **false**. This step is to determine if we want to use the test bucket or prod bucket to store the binary files.

## Set up
**Trust policy for github-action-role**
```json
{
	"Version": "2012-10-17",
	"Statement": [
		{
			"Effect": "Allow",
			"Principal": {
				"Federated": "arn:aws:iam::<aws_account_number>:oidc-provider/token.actions.githubusercontent.com"
			},
			"Action": "sts:AssumeRoleWithWebIdentity",
			"Condition": {
				"StringLike": {
					"token.actions.githubusercontent.com:sub": [
						"repo:valkey/valkey:ref:refs/heads/*",
						"repo:vudiep411/valkey:ref:refs/tags/*",
						"repo:vudiep411/valkey:pull_request"
					],
					"token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
				}
			}
		}
	]
}
```
**Repo Secrets**
`AWS_ROLE_TO_ASSUME`: ARN role created from the prerequisite
`AWS_S3_BUCKET`: Prod bucket name
`AWS_TEST_BUCKET`: Test bucket name

## Results
* This is the [PR](https://github.com/vudiep411/valkey/pull/7) replicated in my env.
* Check the CI build-release for this PR [here](https://github.com/vudiep411/valkey/pull/7/checks)

**AWS Buckets**
We will have 2 buckets. 1 is a test bucket and the other is a prod bucket
Screenshot 2024-12-02 at 2 42 15 PM

## Summary
The PR CI build-release will build the binary files into the `test bucket` to check if these files are correctly build. After that we can merge the PR and whenever we release a new version, this workflow will run and build the binary files into the production bucket.

@roshkhatri

File: .github/workflows/build-release-packages.yml
Comment (vudiep411) [inline .github/workflows/build-release-packages.yml]: Check if whether it is a pull request or it's an actual release (manual or automated) and output true or false. This step purpose is to know if we should use a `test` bucket or a `prod` bucket

File: .github/workflows/build-release-packages.yml
Comment (vudiep411) [inline .github/workflows/build-release-packages.yml:93]: Passing in the either the `test_bucket` or `prod_bucket` to run the reusable workflow.

**Comment by vudiep411** [comment]
## New Changes
1. When a PR is raised, the build-release will run if there are changes to the release workflow or `utils/releasetools`
2. `Get the version` step will output the version that is getting built. eg. 8.0.0, unstable,...
3. `Set bucket name` test will output **true** if the event is a PR else it will output **false**. This step is to determine if we want to use the test bucket or prod bucket to store the binary files.

## Set up
**Trust policy for github-action-role**
```json
{
	"Version": "2012-10-17",
	"Statement": [
		{
			"Effect": "Allow",
			"Principal": {
				"Federated": "arn:aws:iam::<aws_account_number>:oidc-provider/token.actions.githubusercontent.com"
			},
			"Action": "sts:AssumeRoleWithWebIdentity",
			"Condition": {
				"StringLike": {
					"token.actions.githubusercontent.com:sub": [
						"repo:valkey/valkey:ref:refs/heads/*",
						"repo:vudiep411/valkey:ref:refs/tags/*",
						"repo:vudiep411/valkey:pull_request"
					],
					"token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
				}
			}
		}
	]
}
```
**Repo Secrets**
`AWS_ROLE_TO_ASSUME`: ARN role created from the prerequisite
`AWS_S3_BUCKET`: Prod bucket name
`AWS_TEST_BUCKET`: Test bucket name

## Results
* This is the [PR](https://github.com/vudiep411/valkey/pull/7) replicated in my env.
* Check the CI build-release for this PR [here](https://github.com/vudiep411/valkey/pull/7/checks)

**AWS Buckets**
We will have 2 buckets. 1 is a test bucket and the other is a prod bucket
Screenshot 2024-12-02 at 2 42 15 PM

## Summary
The PR CI build-release will build the binary files into the `test bucket` to check if these files are correctly build. After that we can merge the PR and whenever we release a new version, this workflow will run and build the binary files into the production bucket.

@roshkhatri

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml]: Pull request can be on any branch and we would need to take the parent branch.

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml]: Can you move this to a place just before we use it?

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml]:
```suggestion
      - name: Check if we are testing.
        id: check-if-testing
        run: |
          if [[ "${{ github.event_name }}" == "pull_request" ]]; then
            echo "IS_TEST=true" >> $GITHUB_OUTPUT
          else
            echo "IS_TEST=false" >> $GITHUB_OUTPUT
          fi
        shell: bash
```

**Review by roshkhatri — COMMENTED** [review]
LGTM

**Review by roshkhatri — APPROVED** [review]
LGTM! once this PR is approve, we can make the necessary changes to the prerequisite AWS configs and merge this one in. 

Can you also update the top comment to mention all the changes we made here later
such as, adding tests, etc.

**Review by madolson — APPROVED** [review]

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml]:
```suggestion
  push:
```
I just realized that the secrets will not be accessible on a pull request, rather we should do it on a push event

File: .github/workflows/build-release-packages.yml
Comment (roshkhatri) [inline .github/workflows/build-release-packages.yml]: Please add these back and use
`utils/releasetools/build-config.json`

**Pull request merged**
